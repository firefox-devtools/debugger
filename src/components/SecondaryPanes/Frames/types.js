import type { Frame } from "debugger-html";

export type LocalFrame = Frame & {
  library: string,
  source: Source
};
