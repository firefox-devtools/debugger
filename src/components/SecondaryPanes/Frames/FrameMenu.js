// @flow
import { showMenu } from "devtools-launchpad";
import { copyToTheClipboard } from "../../../utils/clipboard";
import type { LocalFrame } from "./types";
import type { ContextMenuItem } from "debugger-html";
import { kebabCase } from "lodash";

const blackboxString = "sourceFooter.blackbox";
const unblackboxString = "sourceFooter.unblackbox";

function formatMenuElement(
  labelString: string,
  click: Function,
  disabled: boolean = false
): ContextMenuItem {
  const label = L10N.getStr(labelString);
  const accesskey = L10N.getStr(`${labelString}.accesskey`);
  const id = `node-menu-${kebabCase(label)}`;
  return {
    id,
    label,
    accesskey,
    disabled,
    click
  };
}

function copySourceElement(url) {
  return formatMenuElement("copySourceUri2", () => copyToTheClipboard(url));
}

function copyStackTraceElement(copyStackTrace) {
  return formatMenuElement("copyStackTrace", () => copyStackTrace());
}

function toggleFrameworkGroupingElement(
  toggleFrameworkGrouping,
  frameworkGroupingOn
) {
  const actionType = frameworkGroupingOn
    ? "framework.disableGrouping"
    : "framework.enableGrouping";

  return formatMenuElement(actionType, () => toggleFrameworkGrouping());
}

function blackBoxSource(source, toggleBlackBox) {
  const toggleBlackBoxString = source.isBlackBoxed
    ? unblackboxString
    : blackboxString;

  return formatMenuElement(toggleBlackBoxString, () => toggleBlackBox(source));
}

export default function FrameMenu(
  frame: LocalFrame,
  frameworkGroupingOn: boolean,
  callbacks: Object,
  event: SyntheticKeyboardEvent
) {
  event.stopPropagation();
  event.preventDefault();

  const menuOptions = [];

  const source = frame.source;

  const toggleFrameworkElement = toggleFrameworkGroupingElement(
    callbacks.toggleFrameworkGrouping,
    frameworkGroupingOn
  );
  menuOptions.push(toggleFrameworkElement);

  if (source) {
    const copySourceUri2 = copySourceElement(source.url);
    menuOptions.push(copySourceUri2);
    menuOptions.push(blackBoxSource(source, callbacks.toggleBlackBox));
  }

  const copyStackTraceItem = copyStackTraceElement(callbacks.copyStackTrace);

  menuOptions.push(copyStackTraceItem);

  showMenu(event, menuOptions);
}
