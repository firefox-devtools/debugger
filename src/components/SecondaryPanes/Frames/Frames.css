.frames ul {
  list-style: none;
  margin: 0;
  padding: 0;
}

.frames ul li {
  padding: 7px 10px 7px 21px;
  overflow: hidden;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  align-items: center;
  margin: 0;
}

.frames ul li * {
  -moz-user-select: none;
  user-select: none;
}

.frames .location {
  font-weight: lighter;
  display: flex;
  justify-content: space-between;
  flex-direction: row;
  align-items: center;
  margin: 0;
}

.theme-light .frames .location,
.theme-firebug .frames .location {
  color: var(--theme-comment);
}

:root.theme-dark .frames .location {
  color: var(--theme-body-color);
  opacity: 0.6;
}

.frames .title {
  text-overflow: ellipsis;
  overflow: hidden;
  margin-right: 1em;
}

.frames ul li:hover,
.frames ul li:focus {
  background-color: var(--theme-toolbar-background-alt);
  outline: none;
}

.theme-dark .frames ul li:focus {
  background-color: var(--theme-tab-toolbar-background);
}

.frames ul li.selected {
  background-color: var(--theme-selection-background);
  color: white;
}

.frames ul li.selected i.annotation-logo svg path {
  fill: white;
}

:root.theme-light .frames ul li.selected .location,
:root.theme-firebug .frames ul li.selected .location,
:root.theme-dark .frames ul li.selected .location {
  color: white;
}

.show-more {
  text-align: center;
  padding: 8px 0px;
  margin: 7px 10px 7px 7px;
  border: 1px solid var(--theme-splitter-color);
  background-color: var(--theme-tab-toolbar-background);
}

.show-more:hover {
  background-color: var(--theme-toolbar-background-hover);
}

.annotation-logo {
  width: 12px;
  margin-left: 3px;
  line-height: 8px;
}

:root.theme-dark .annotation-logo svg path {
  fill: var(--theme-highlight-blue);
}
