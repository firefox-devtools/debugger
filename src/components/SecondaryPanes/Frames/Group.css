.frames ul .frames-group .group,
.frames ul .frames-group .group .location {
  font-weight: 500;
  cursor: default;
}

.frames ul .frames-group.expanded .group,
.frames ul .frames-group.expanded .group .location {
  color: var(--theme-highlight-blue);
}

.frames ul .frames-group.expanded .react path {
  fill: var(--theme-highlight-blue);
}

.frames ul .frames-group .frames-list li {
  padding-left: 30px;
}

.frames ul .frames-group .frames-list {
  border-top: 1px solid var(--theme-splitter-color);
  border-bottom: 1px solid var(--theme-splitter-color);
}
