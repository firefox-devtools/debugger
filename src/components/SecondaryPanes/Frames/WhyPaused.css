.why-paused {
  background-color: var(--theme-body-background);
  color: var(--theme-body-color);
  padding: 10px 10px 10px 20px;
  white-space: normal;
  opacity: 0.6;
  font-size: 12px;
  flex: 0 1 auto;
  text-align: center;
  font-style: italic;
  font-weight: 300;
  cursor: default;
}

.theme-dark .secondary-panes .why-paused {
  color: white;
}

.why-paused .message {
  font-size: 10px;
}

.why-paused .message.warning {
  font-size: 10px;
  color: var(--theme-graphs-red);
  font-weight: bold;
  font-style: normal;
}
