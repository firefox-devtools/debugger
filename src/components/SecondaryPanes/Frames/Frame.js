/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow
import React, { Component } from "react";
import PropTypes from "prop-types";

import classNames from "classnames";
import Svg from "../../shared/Svg";

import { formatDisplayName } from "../../../utils/pause/frames";
import { getFilename, getFileURL } from "../../../utils/source";
import FrameMenu from "./FrameMenu";

import type { Frame } from "../../../types";
import type { LocalFrame } from "./types";

type FrameTitleProps = {
  frame: Frame,
  options: Object,
  l10n: Object
};

function FrameTitle({ frame, options = {}, l10n }: FrameTitleProps) {
  const displayName = formatDisplayName(frame, options, l10n);
  return <span className="title">{displayName}</span>;
}

type FrameLocationProps = { frame: LocalFrame, displayFullUrl: boolean };

function FrameLocation({ frame, displayFullUrl = false }: FrameLocationProps) {
  if (!frame.source) {
    return null;
  }

  if (frame.library) {
    return (
      <span className="location">
        {frame.library}
        <Svg name={frame.library.toLowerCase()} className="annotation-logo" />
      </span>
    );
  }

  const { location, source } = frame;
  const filename = displayFullUrl
    ? getFileURL(source, false)
    : getFilename(source);

  return (
    <span className="location">
      <span className="filename">{filename}</span>:
      <span className="line">{location.line}</span>
    </span>
  );
}

FrameLocation.displayName = "FrameLocation";

type FrameComponentProps = {
  frame: LocalFrame,
  selectedFrame: LocalFrame,
  copyStackTrace: Function,
  toggleFrameworkGrouping: Function,
  selectFrame: Function,
  frameworkGroupingOn: boolean,
  hideLocation: boolean,
  shouldMapDisplayName: boolean,
  toggleBlackBox: Function,
  displayFullUrl: boolean,
  getFrameTitle?: string => string,
  disableContextMenu: boolean
};

export default class FrameComponent extends Component<FrameComponentProps> {
  static defaultProps = {
    hideLocation: false,
    shouldMapDisplayName: true,
    disableContextMenu: false
  };

  onContextMenu(event: SyntheticMouseEvent<HTMLElement>) {
    const {
      frame,
      copyStackTrace,
      toggleFrameworkGrouping,
      toggleBlackBox,
      frameworkGroupingOn
    } = this.props;
    FrameMenu(
      frame,
      frameworkGroupingOn,
      { copyStackTrace, toggleFrameworkGrouping, toggleBlackBox },
      event
    );
  }

  onMouseDown(
    e: SyntheticMouseEvent<HTMLElement>,
    frame: Frame,
    selectedFrame: Frame
  ) {
    if (e.button !== 0) {
      return;
    }
    this.props.selectFrame(frame);
  }

  onKeyUp(
    event: SyntheticKeyboardEvent<HTMLElement>,
    frame: Frame,
    selectedFrame: Frame
  ) {
    if (event.key != "Enter") {
      return;
    }
    this.props.selectFrame(frame);
  }

  render() {
    const {
      frame,
      selectedFrame,
      hideLocation,
      shouldMapDisplayName,
      displayFullUrl,
      getFrameTitle,
      disableContextMenu
    } = this.props;
    const { l10n } = this.context;

    const className = classNames("frame", {
      selected: selectedFrame && selectedFrame.id === frame.id
    });

    const title = getFrameTitle
      ? getFrameTitle(
          `${getFileURL(frame.source, false)}:${frame.location.line}`
        )
      : undefined;

    const tabChar = "\t";
    const newLineChar = "\n";

    return (
      <li
        key={frame.id}
        className={className}
        onMouseDown={e => this.onMouseDown(e, frame, selectedFrame)}
        onKeyUp={e => this.onKeyUp(e, frame, selectedFrame)}
        onContextMenu={disableContextMenu ? null : e => this.onContextMenu(e)}
        tabIndex={0}
        title={title}
      >
        {tabChar}
        <FrameTitle
          frame={frame}
          options={{ shouldMapDisplayName }}
          l10n={l10n}
        />
        {!hideLocation && " "}
        {!hideLocation && (
          <FrameLocation frame={frame} displayFullUrl={displayFullUrl} />
        )}
        {newLineChar}
      </li>
    );
  }
}

FrameComponent.displayName = "Frame";
FrameComponent.contextTypes = { l10n: PropTypes.object };
