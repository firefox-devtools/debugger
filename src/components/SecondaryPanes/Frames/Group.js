/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow
import React, { Component } from "react";
import classNames from "classnames";
import Svg from "../../shared/Svg";
import { formatDisplayName, getLibraryFromUrl } from "../../../utils/frame";
import FrameMenu from "./FrameMenu";

import "./Group.css";

import FrameComponent from "./Frame";

import type { LocalFrame } from "./types";
import type { Frame } from "../../../types";
import Badge from "../../shared/Badge";

type FrameLocationProps = { frame: Frame };
function FrameLocation({ frame }: FrameLocationProps) {
  const library = getLibraryFromUrl(frame);
  if (!library) {
    return null;
  }

  return (
    <div className="location">
      {library}
      <Svg name={library.toLowerCase()} className="annotation-logo" />
    </div>
  );
}

FrameLocation.displayName = "FrameLocation";

type Props = {
  group: LocalFrame[],
  selectedFrame: LocalFrame,
  selectFrame: Function,
  toggleFrameworkGrouping: Function,
  copyStackTrace: Function,
  toggleBlackBox: Function,
  frameworkGroupingOn: boolean
};

type State = {
  expanded: boolean
};

export default class Group extends Component<Props, State> {
  toggleFrames: Function;

  constructor(...args: any[]) {
    super(...args);
    this.state = { expanded: false };
  }

  onContextMenu(event: SyntheticKeyboardEvent<HTMLElement>) {
    const {
      group,
      copyStackTrace,
      toggleFrameworkGrouping,
      toggleBlackBox,
      frameworkGroupingOn
    } = this.props;
    const frame = group[0];
    FrameMenu(
      frame,
      frameworkGroupingOn,
      { copyStackTrace, toggleFrameworkGrouping, toggleBlackBox },
      event
    );
  }

  toggleFrames = () => {
    this.setState({ expanded: !this.state.expanded });
  };

  renderFrames() {
    const {
      group,
      selectFrame,
      selectedFrame,
      toggleFrameworkGrouping,
      frameworkGroupingOn,
      toggleBlackBox,
      copyStackTrace
    } = this.props;

    const { expanded } = this.state;
    if (!expanded) {
      return null;
    }

    return (
      <div className="frames-list">
        {group.map(frame => (
          <FrameComponent
            copyStackTrace={copyStackTrace}
            frame={frame}
            frameworkGroupingOn={frameworkGroupingOn}
            hideLocation={true}
            key={frame.id}
            selectedFrame={selectedFrame}
            selectFrame={selectFrame}
            shouldMapDisplayName={false}
            toggleBlackBox={toggleBlackBox}
            toggleFrameworkGrouping={toggleFrameworkGrouping}
          />
        ))}
      </div>
    );
  }

  renderDescription() {
    const frame = this.props.group[0];
    const displayName = formatDisplayName(frame);
    return (
      <li
        key={frame.id}
        className={classNames("group")}
        onClick={this.toggleFrames}
        tabIndex={0}
      >
        <div className="d-flex align-items-center min-width-0">
          <div className="title">{displayName}</div>
          <Badge>{this.props.group.length}</Badge>
        </div>
        <FrameLocation frame={frame} />
      </li>
    );
  }

  render() {
    const { expanded } = this.state;
    return (
      <div
        className={classNames("frames-group", { expanded })}
        onContextMenu={e => this.onContextMenu(e)}
      >
        {this.renderDescription()}
        {this.renderFrames()}
      </div>
    );
  }
}

Group.displayName = "Group";
