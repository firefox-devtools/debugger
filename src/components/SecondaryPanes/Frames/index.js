/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */

// @flow

import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";

import type { Frame, Why } from "../../../types";

import FrameComponent from "./Frame";
import Group from "./Group";

import renderWhyPaused from "./WhyPaused";

import actions from "../../../actions";
import { collapseFrames, formatCopyName } from "../../../utils/pause/frames";
import { copyToTheClipboard } from "../../../utils/clipboard";

import {
  getFrameworkGroupingState,
  getSelectedFrame,
  isPaused as getIsPaused,
  getCallStackFrames,
  getPauseReason
} from "../../../selectors";

import type { LocalFrame } from "./types";

import "./Frames.css";

const NUM_FRAMES_SHOWN = 7;

type Props = {
  frames: Array<Frame>,
  frameworkGroupingOn: boolean,
  selectedFrame: Object,
  why: Why,
  selectFrame: Function,
  toggleBlackBox: Function,
  toggleFrameworkGrouping: Function,
  disableFrameTruncate: boolean,
  disableContextMenu: boolean,
  displayFullUrl: boolean,
  getFrameTitle?: string => string
};

type State = {
  showAllFrames: boolean
};

class Frames extends Component<Props, State> {
  renderFrames: Function;
  toggleFramesDisplay: Function;
  truncateFrames: Function;
  copyStackTrace: Function;
  toggleFrameworkGrouping: Function;
  renderToggleButton: Function;

  constructor(props: Props) {
    super(props);

    this.state = {
      showAllFrames: !!props.disableFrameTruncate
    };
  }

  shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
    const { frames, selectedFrame, frameworkGroupingOn } = this.props;
    const { showAllFrames } = this.state;
    return (
      frames !== nextProps.frames ||
      selectedFrame !== nextProps.selectedFrame ||
      showAllFrames !== nextState.showAllFrames ||
      frameworkGroupingOn !== nextProps.frameworkGroupingOn
    );
  }

  toggleFramesDisplay = (): void => {
    this.setState(prevState => ({
      showAllFrames: !prevState.showAllFrames
    }));
  };

  collapseFrames(frames: Array<Frame>) {
    const { frameworkGroupingOn } = this.props;
    if (!frameworkGroupingOn) {
      return frames;
    }

    return collapseFrames(frames);
  }

  truncateFrames(frames: Array<Frame>): Array<Frame> {
    const numFramesToShow = this.state.showAllFrames
      ? frames.length
      : NUM_FRAMES_SHOWN;

    return frames.slice(0, numFramesToShow);
  }

  copyStackTrace = () => {
    const { frames } = this.props;
    const { l10n } = this.context;
    const framesToCopy = frames.map(f => formatCopyName(f, l10n)).join("\n");
    copyToTheClipboard(framesToCopy);
  };

  toggleFrameworkGrouping = () => {
    const { toggleFrameworkGrouping, frameworkGroupingOn } = this.props;
    toggleFrameworkGrouping(!frameworkGroupingOn);
  };

  renderFrames(frames: LocalFrame[]) {
    const {
      selectFrame,
      selectedFrame,
      toggleBlackBox,
      frameworkGroupingOn,
      displayFullUrl,
      getFrameTitle,
      disableContextMenu
    } = this.props;

    const framesOrGroups = this.truncateFrames(this.collapseFrames(frames));
    type FrameOrGroup = LocalFrame | LocalFrame[];

    return (
      <ul>
        {framesOrGroups.map(
          (frameOrGroup: FrameOrGroup) =>
            frameOrGroup.id ? (
              <FrameComponent
                frame={frameOrGroup}
                toggleFrameworkGrouping={this.toggleFrameworkGrouping}
                copyStackTrace={this.copyStackTrace}
                frameworkGroupingOn={frameworkGroupingOn}
                selectFrame={selectFrame}
                selectedFrame={selectedFrame}
                toggleBlackBox={toggleBlackBox}
                key={String(frameOrGroup.id)}
                displayFullUrl={displayFullUrl}
                getFrameTitle={getFrameTitle}
                disableContextMenu={disableContextMenu}
              />
            ) : (
              <Group
                group={frameOrGroup}
                toggleFrameworkGrouping={this.toggleFrameworkGrouping}
                copyStackTrace={this.copyStackTrace}
                frameworkGroupingOn={frameworkGroupingOn}
                selectFrame={selectFrame}
                selectedFrame={selectedFrame}
                toggleBlackBox={toggleBlackBox}
                key={frameOrGroup[0].id}
                displayFullUrl={displayFullUrl}
                getFrameTitle={getFrameTitle}
                disableContextMenu={disableContextMenu}
              />
            )
        )}
      </ul>
    );
  }

  renderToggleButton(frames: LocalFrame[]) {
    const { l10n } = this.context;
    const buttonMessage = this.state.showAllFrames
      ? l10n.getStr("callStack.collapse")
      : l10n.getStr("callStack.expand");

    frames = this.collapseFrames(frames);
    if (frames.length <= NUM_FRAMES_SHOWN) {
      return null;
    }

    return (
      <div className="show-more-container">
        <button className="show-more" onClick={this.toggleFramesDisplay}>
          {buttonMessage}
        </button>
      </div>
    );
  }

  render() {
    const { frames, disableFrameTruncate, why } = this.props;

    if (!frames) {
      return (
        <div className="pane frames">
          <div className="pane-info empty">
            {L10N.getStr("callStack.notPaused")}
          </div>
        </div>
      );
    }

    return (
      <div className="pane frames">
        {this.renderFrames(frames)}
        {renderWhyPaused(why)}
        {disableFrameTruncate ? null : this.renderToggleButton(frames)}
      </div>
    );
  }
}

Frames.contextTypes = { l10n: PropTypes.object };

const mapStateToProps = state => ({
  frames: getCallStackFrames(state),
  why: getPauseReason(state),
  frameworkGroupingOn: getFrameworkGroupingState(state),
  selectedFrame: getSelectedFrame(state),
  pause: getIsPaused(state)
});

export default connect(
  mapStateToProps,
  {
    selectFrame: actions.selectFrame,
    toggleBlackBox: actions.toggleBlackBox,
    toggleFrameworkGrouping: actions.toggleFrameworkGrouping,
    disableFrameTruncate: false,
    disableContextMenu: false,
    displayFullUrl: false
  }
)(Frames);

// Export the non-connected component in order to use it outside of the debugger
// panel (e.g. console, netmonitor, â€¦).
export { Frames };
