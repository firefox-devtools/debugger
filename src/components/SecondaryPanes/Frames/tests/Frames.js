import React from "react";
import { shallow } from "enzyme";
import { Map } from "immutable";
import Frames, { getAndProcessFrames } from "../index.js";

function render(overrides = {}) {
  const defaultProps = {
    frames: null,
    selectedFrame: null,
    frameworkGroupingOn: false,
    toggleFrameworkGrouping: jest.fn(),
    contextTypes: {},
    selectFrame: jest.fn(),
    toggleBlackBox: jest.fn()
  };

  const props = Object.assign({}, defaultProps, overrides);
  const component = shallow(<Frames.WrappedComponent {...props} />);

  return component;
}

describe("Frames", () => {
  describe("Supports different number of frames", () => {
    it("empty frames", () => {
      const component = render();
      expect(component).toMatchSnapshot();
      expect(component.find(".show-more").exists()).toBeFalsy();
    });

    it("one frame", () => {
      const frames = [{ id: 1 }];
      const selectedFrame = frames[0];
      const component = render({ frames, selectedFrame });

      expect(component.find(".show-more").exists()).toBeFalsy();
      expect(component).toMatchSnapshot();
    });

    it("toggling the show more button", () => {
      const frames = [
        { id: 1 },
        { id: 2 },
        { id: 3 },
        { id: 4 },
        { id: 5 },
        { id: 6 },
        { id: 7 },
        { id: 8 },
        { id: 9 },
        { id: 10 }
      ];

      const selectedFrame = frames[0];
      const component = render({ selectedFrame, frames });

      const getToggleBtn = () => component.find(".show-more");
      const getFrames = () => component.find("FrameComponent");

      expect(getToggleBtn().text()).toEqual("Expand rows");
      expect(getFrames().length).toEqual(7);

      getToggleBtn().simulate("click");
      expect(getToggleBtn().text()).toEqual("Collapse rows");
      expect(getFrames().length).toEqual(10);
      expect(component).toMatchSnapshot();
    });
  });

  describe("Blackboxed Frames", () => {
    it("filters blackboxed frames", () => {
      const frames = [
        { id: 1, location: { sourceId: "1" } },
        { id: 2, location: { sourceId: "2" } },
        { id: 3, location: { sourceId: "1" } },
        { id: 8, location: { sourceId: "2" } }
      ];

      const sources = Map({
        1: Map({}),
        2: Map({ isBlackBoxed: true })
      });

      const processedFrames = getAndProcessFrames(frames, sources);
      const selectedFrame = frames[0];

      const component = render({
        frames: processedFrames,
        frameworkGroupingOn: false,
        selectedFrame
      });

      expect(component.find("FrameComponent").length).toEqual(2);
      expect(component).toMatchSnapshot();
    });
  });

  describe("Library Frames", () => {
    it("toggling framework frames", () => {
      const frames = [
        { id: 1 },
        { id: 2, library: "back" },
        { id: 3, library: "back" },
        { id: 8 }
      ];

      const selectedFrame = frames[0];
      const frameworkGroupingOn = false;
      const component = render({ frames, frameworkGroupingOn, selectedFrame });

      expect(component.find("FrameComponent").length).toEqual(4);
      expect(component).toMatchSnapshot();

      component.setProps({ frameworkGroupingOn: true });

      expect(component.find("FrameComponent").length).toEqual(2);
      expect(component).toMatchSnapshot();
    });
  });
});
