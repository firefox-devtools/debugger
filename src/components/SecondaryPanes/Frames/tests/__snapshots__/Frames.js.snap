// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frames Blackboxed Frames filters blackboxed frames 1`] = `
<div
  className="pane frames"
>
  <ul>
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 1,
          "location": Object {
            "sourceId": "1",
          },
          "source": Object {},
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="1"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
          "location": Object {
            "sourceId": "1",
          },
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 3,
          "location": Object {
            "sourceId": "1",
          },
          "source": Object {},
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="3"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
          "location": Object {
            "sourceId": "1",
          },
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
  </ul>
</div>
`;

exports[`Frames Library Frames toggling framework frames 1`] = `
<div
  className="pane frames"
>
  <ul>
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 1,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="1"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 2,
          "library": "back",
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="2"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 3,
          "library": "back",
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="3"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 8,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="8"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
  </ul>
</div>
`;

exports[`Frames Library Frames toggling framework frames 2`] = `
<div
  className="pane frames"
>
  <ul>
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 1,
        }
      }
      frameworkGroupingOn={true}
      hideLocation={false}
      key="1"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Group
      copyStackTrace={[Function]}
      frameworkGroupingOn={true}
      group={
        Array [
          Object {
            "id": 2,
            "library": "back",
          },
          Object {
            "id": 3,
            "library": "back",
          },
        ]
      }
      key="2"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 8,
        }
      }
      frameworkGroupingOn={true}
      hideLocation={false}
      key="8"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
  </ul>
</div>
`;

exports[`Frames Supports different number of frames empty frames 1`] = `
<div
  className="pane frames"
>
  <div
    className="pane-info empty"
  >
    Not paused
  </div>
</div>
`;

exports[`Frames Supports different number of frames one frame 1`] = `
<div
  className="pane frames"
>
  <ul>
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 1,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="1"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
  </ul>
</div>
`;

exports[`Frames Supports different number of frames toggling the show more button 1`] = `
<div
  className="pane frames"
>
  <ul>
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 1,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="1"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 2,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="2"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 3,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="3"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 4,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="4"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 5,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="5"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 6,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="6"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 7,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="7"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 8,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="8"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 9,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="9"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
    <Frame
      copyStackTrace={[Function]}
      frame={
        Object {
          "id": 10,
        }
      }
      frameworkGroupingOn={false}
      hideLocation={false}
      key="10"
      selectFrame={[Function]}
      selectedFrame={
        Object {
          "id": 1,
        }
      }
      shouldMapDisplayName={true}
      toggleBlackBox={[Function]}
      toggleFrameworkGrouping={[Function]}
    />
  </ul>
  <div
    className="show-more"
    onClick={[Function]}
  >
    Collapse rows
  </div>
</div>
`;
